### DISCUSSION TITLE
IntelliJ Plugin is missing

### DISCUSSION THREAD

**Discussion by gmitch215** [korlibs/korge#2296] [Q&A] [answered]
Woke up to find that the IntelliJ plugin no longer exists.

Can't search for it, and the [direct link](https://plugins.jetbrains.com/plugin/9676-korge) reports a 404.

![image](https://github.com/user-attachments/assets/011c266b-075b-4fe9-af21-b9be7cd16beb)

**Comment by soywiz** [comment, marked as answer]
Korge plugin was removed from the marketplace as it was outdated.

There are three options now:
* Use https://forge.korge.org/
* Use Intellij without the Korge plugin. It is not required to use Korge with gradle
* If you are using an old intellij version compatible with the plugin: https://github.com/korlibs/korge-intellij-plugin/releases/tag/v4.0.4 here is a zip file you can install in IntelliJ

**Reply by wbehrens-on-gh** [reply]
As others have expressed here I think this is a step backwards. What originally drew me to korge was how simple it was to setup with just an existing Intellij installation. I personally hope this decision can get reversed in the future.

**Reply by soywiz** [reply]
If a team is formed, and they have bandwidth to support it and keeping up to date with deprecations and changes and decide it is worth to do it, of course. I'm not against it.

Also have in mind:

* Forge is installed with one command and two clicks.
* Korge can be used in intellij normally, as it uses gradle, and it is a plain kotlin library. Gradle tasks are there.

**Comment by gmitch215** [comment]
If the plugin wasn't necessary for development, what was its original purpose? Is this purpose still relevant?

**Reply by soywiz** [reply]
Project templates, store and documentation acces inside the IDE, intents for color choosing among others, resource explorer and hot reloading preview inside the IDE, etc.
All that is optional to use KorGE, as it is a plain library and a Gradle plugin.

KorGE Forge is like an IntelliJ with the plugin. It is done to provide a version of the IntelliJ platform that is compatible with the code while providing a better experience by removing functionality unrelated to Kotlin. For KorGE development, Forge is the supported and recommended way, but optional.

**Reply by FyiurAmron** [reply]
Frankly, that's a really bad choice to take. All the major game engines provide IDE plugins _even if_ they have native IDEs or editors provided. The reason is obvious, to streamline the dev's workflow and to allow them to use the tools they have - including configs, other plugins, other projects etc. It also makes getting support for IDE trivial, in case of bugs and problems. A custom forked platform means literally no support in the end.

> KorGE Forge is like an IntelliJ with the plugin.

If so, why maintain an IntelliJ platform conversion instead of a plugin? It makes no sense.

> It is done to provide a version of the IntelliJ platform that is compatible with the code while providing a better experience by removing functionality unrelated to Kotlin.

You're telling me you know better than me what functionalities I need or want to have in my IDE and what will "provid[e] a better experience" for me? A bold statement. I've coded for 30 years, I think I know a bit about my own experience already.

>For KorGE development, Forge is the supported and recommended way, but optional.

So, basically, you're saying that instead of using our regular IDEs + a dedicated plugin, like for UE, Unity, Godot or whatnot - those who don't want to use an entirely separate platform for KorGE dev are left only with base IDE and no additional KorGE support _or_ a legacy IDEA version?

I do see some value for some people in providing a separate IDEA IDE platform (even though that's rather uncommon and somewhat unnecessary in this case TBH)... but abandoning a core IDEA plugin? That doesn't seem to be a smart move, I'd say.

**Reply by FyiurAmron** [reply]
![image](https://github.com/user-attachments/assets/b1a25204-4cd8-4f74-a4cb-3a9a3af3224c)
^ believe it or not, but this is the "future" you're actively creating for people who wanted to try KorGE :)

**Reply by FyiurAmron** [reply]
also, FWIW @soywiz : https://korge.org/features/quicksetup is very outdated, some links are dead and things mentioned there don't really reflect current reality AFAIK.

**Reply by gmitch215** [reply]
@FyiurAmron I don't blame him - the entire ecosystem is essentially only managed and maintained by him. Given the size of this project, it's amazing it's even up-to-date.

He said he'd return from a hiatus around this time, though I haven't heard from him in a while.
I think he just did what he thought was best for his project.

**Reply by FyiurAmron** [reply]
@gmitch215 I'm not blaming him, because this is not about blame. And yes, I also do consider this choice (switch from plugin to IDEA fork) an attempt to improve, even if arguably failed one. However, it's still about taking a step back and looking at the engineering problems & solutions and whether they work or not. I present the case why they don't currently. IMVHO, Forge just doesn't make sense in its current form, precisely for the reason I stated - maintaining it as an IDEA platform fork is an order of magnitude harder both for devs and users than providing a plugin. The argument "but at least it can stay that way indefinitely instead of a plugin which will get incompatible" doesn't hold, because the IDEA fork has to be updated anyway to work (code rot happens anyway).

Also, it's a ~2 GiB and completely unneeded installation, not to mention the added moving parts here and the failures caused by them (see the screenshot - this is from my actual attempt to use Forge, dunno if it's a "hard" bug [likely] or a quirk of my system [unlikely, but possible] - still, trying to use it just for an example project just fails due to internal IntelliJ exceptions being thrown around).

And yes, I say that precisely because I'm doing exactly this kind of stuff in my day job, and has been doing it for a while. Choices like this backfire, usually quite quickly (months or years, not decades). Hell, I wouldn't mind updating that plugin if I'd to stick with KorGE (at least I'll up my game with IntelliJ plugins again, haven't worked with them in some years) - no way I'm touching a platform fork (learning how to do it is useless for me as it makes no sense for most cases, precisely for the reasons described above - for Java/Kotlin there already _is_ a verified, stable and robust alternative, and doing so is by itself a tedious coding exercise, but nothing more).

A side note: the fact that this is a single-person project is not a viable "line of defense" for deficiencies in neither FOSS nor commercial closed-source in XXI century. If he wanted or needed more developers, he could ask around, and given enough effort they would come. There is precisely 0 _need_ for the project to be 1-person, unless the project is obscure _and_ mostly useless (which I really don't believe is the case here - it's reasonably useful and there is a community of actual _coders_ around KorGE). Keeping something mostly 1-person is usually a decision of the core dev based on his personality or project management views and approach to team processes, not dire necessity. Also, it's unsustainable in the long run for incubating projects (which KorGE, arguably, is).

**Reply by gmitch215** [reply]
@FyiurAmron Everything you said is correct. I'm sure he had a reason that may not be obvious, or he just did it to gain experience for himself.

There are other members in the korlibs organization, so he's trying to pull help - most of them are inactive or don't work on the main features. There are a million explanations that are reasonable, and a million that are not; one or multiple could be true.

All I'm saying is that he'll do what he thinks is best for his game engine. One good reason for switching to an IDE is implementing a visual editor (which is probably the most likely answer long-term) along with providing more functionality that JetBrains plugins don't have the possibility of doing.

**Reply by FyiurAmron** [reply]
> All I'm saying is that he'll do what he thinks is best for his game engine.

This is what I fear the most. It's literally how great projects with narrow leadership channel fail.
